File: app/src/main/java/com/example/notes/NoteFragment.java
package com.example.notes;

import static com.example.notes.MainActivity.note_text_color;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.Dialog;
import android.content.DialogInterface;
import android.content.res.Configuration;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.PopupMenu;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;


public class NoteFragment extends Fragment {
    //  private OnDialogListener ColorListener;
    static final String SELECTED_NOTE = "note";
    static int PDateTime;
    //  public String mListener;
    boolean title_was_changed;
    private int text_color;
    private Note note;


    public NoteFragment() {
        // Required empty public constructor
    }

    public static NoteFragment newInstance(Note note) {
        NoteFragment fragment = new NoteFragment();
        Bundle args = new Bundle();
        args.putParcelable(SELECTED_NOTE, note);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (savedInstanceState != null)
            requireActivity().getSupportFragmentManager().popBackStack();

    }
    //Call onActivity Create method

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        setHasOptionsMenu(true);
        return inflater.inflate(R.layout.fragment_note, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        Bundle arguments = getArguments();

        TextView time_date_alarm = view.findViewById(R.id.time_date_alarm_view);
        TextView tvDescription = view.findViewById(R.id.tvDescription);
        title_was_changed = false;
        if (arguments != null) {
            note = arguments.getParcelable(SELECTED_NOTE);
            if (note == null)
                note = Note.getNote(0);
            TextView tvTitle = view.findViewById(R.id.tvTitle);
            tvTitle.setText(note.getTitle());
            tvTitle.setTextColor(getResources().getColor(note_text_color));
            tvTitle.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                }

                @Override
                public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                    note.setTitle(charSequence.toString());
                    title_was_changed = true;
                    TextView rt = getActivity().findViewById(note.getId());
                    if (rt == null) {
                        int k = 1;
                        rt = getActivity().findViewById(k);
                    }
                    if ((PDateTime == 0) && (rt != null) && (note != null))
                        rt.setText(note.getTitle());
                }

                @Override
                public void afterTextChanged(Editable editable) {
                }
            });


            time_date_alarm.setText(note.getTimeDateAlarm());
            tvDescription.setText(note.getDescription());
            tvDescription.setTextColor(getResources().getColor(note_text_color));
            tvDescription.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                }

                @Override
                public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                    note.setDescription(charSequence.toString());
                }

                @Override
                public void afterTextChanged(Editable editable) {
                }
            });
            time_date_alarm.setOnLongClickListener(view1 -> {
                initDateTimePopupMenu(time_date_alarm);

                return true;
            });
            Button buttonBack = view.findViewById(R.id.btnBack);
            if (buttonBack != null)
                buttonBack.setOnClickListener(view1 -> {
                    if (title_was_changed) {
                        TextView rt = getActivity().findViewById(note.getId());
                        rt.setText(note.getTitle());
                    }
                    requireActivity().getSupportFragmentManager().popBackStack();
                });
        }
    }

    private void initDateTimePopupMenu(TextView view) {
        Activity activity = requireActivity();
        PopupMenu popupMenu = new PopupMenu(activity, view);
        activity.getMenuInflater().inflate(R.menu.popup_menu, popupMenu.getMenu());
        MenuItem item = popupMenu.getMenu().findItem(R.id.action_popup_delete);
        if (item != null) {
            item.setVisible(false);
        }
        popupMenu.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem menuItem) {
                if (menuItem.getItemId() == R.id.action_popup_correct) {
                    showCorrectDateTime(note);
                }
                return true;
            }
        });
        popupMenu.show();
    }

    private void showCorrectDateTime(Note note) {

        this.note = note;
        if (isLandscape()) {
            showLandCorrectDateTime(note);
        } else {
            showPortCorrectDateTime(note);
        }
    }

    private void showPortCorrectDateTime(Note note) {
        ParentDateTimeFragment parentDateTimeFragment = ParentDateTimeFragment.newInstance(note);
        FragmentManager fragmentManager =
                requireActivity().getSupportFragmentManager();
        FragmentTransaction fragmentTransaction =
                fragmentManager.beginTransaction();
        fragmentTransaction.replace(R.id.notes_container, parentDateTimeFragment, "PARENT_FRAGMENT");
        fragmentTransaction.addToBackStack("");
        fragmentTransaction.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE);
        fragmentTransaction.commit();
    }

    private void showLandCorrectDateTime(Note note) {
        ParentDateTimeFragment parentDateTimeFragment = ParentDateTimeFragment.newInstance(note);
        FragmentManager fragmentManager =
                requireActivity().getSupportFragmentManager();
        FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
        fragmentTransaction.replace(R.id.note_container, parentDateTimeFragment); // замена  фрагмента
        fragmentTransaction.addToBackStack("");
        fragmentTransaction.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE);
        fragmentTransaction.commit();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, @NonNull MenuInflater inflater) {

        Fragment CurrentFragment = requireActivity().getSupportFragmentManager().findFragmentByTag("NOTE_FRAGMENT");

        if ((!isLandscape()) && (menu != null) &&
                (CurrentFragment != null && CurrentFragment.isVisible())) {
            menu.clear();
            inflater.inflate(R.menu.note_fragment_menu, menu);
        }
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        int id = item.getItemId();

        switch (id) {
            case R.id.send_note:
                // Сделать поиск
                createOneButtonAlertDialog("Сделать отослать заметку");
                return true;
            case R.id.add_photo:
                createOneButtonAlertDialog("Сделать добавить фото");
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private boolean isLandscape() {
        return getResources().getConfiguration().orientation
                == Configuration.ORIENTATION_LANDSCAPE;
    }

    private void createOneButtonAlertDialog(String title_window) {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        builder.setTitle(title_window)
                .setMessage(R.string.version)
                .setPositiveButton(R.string.ok, new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int which) {
                    }
                });
        Dialog d = builder.show();
        int textViewId = d.getContext().getResources().getIdentifier("android:id/alertTitle", null, null);
        TextView tv = d.findViewById(textViewId);
        tv.setTextColor(getResources().getColor(R.color.set_text_toast));
        tv.setBackgroundColor(getResources().getColor(R.color.teal_200));
        textViewId = d.getContext().getResources().getIdentifier("android:id/message", null, null);
        tv = d.findViewById(textViewId);
        tv.setTextColor(getResources().getColor(android.R.color.holo_red_dark));
        textViewId = d.getContext().getResources().getIdentifier("android:id/button1", null, null);
        Button b = d.findViewById(textViewId);
        b.setTextColor(getResources().getColor(android.R.color.holo_green_dark));
        b.setBackgroundColor(getResources().getColor(R.color.color_time_date));
    }
}

File: app/src/main/java/com/example/notes/NoteFragment.kt
package com.example.notes

import com.example.notes.ListFragmentV2.Companion.newInstance
import com.example.notes.DatePickerFragment.Companion.newInstance
import com.example.notes.Adapter.SectionPagerAdapter.addFragment
import android.os.Parcelable
import androidx.annotation.RequiresApi
import android.os.Build
import android.os.Parcel
import android.os.Parcelable.Creator
import android.annotation.SuppressLint
import android.os.Bundle
import com.example.notes.R
import android.widget.TextView
import com.example.notes.NoteFragment
import com.example.notes.MainActivity
import android.text.TextWatcher
import com.example.notes.CardSourceImpl
import com.example.notes.CardData
import com.google.gson.GsonBuilder
import com.example.notes.ListFragmentV2
import android.text.Editable
import android.view.View.OnLongClickListener
import android.app.Activity
import android.app.AlertDialog
import android.app.Dialog
import com.example.notes.ParentDateTimeFragment
import android.content.DialogInterface
import android.content.res.Configuration
import android.view.*
import com.example.notes.NotesFragment
import com.google.android.material.floatingactionbutton.FloatingActionButton
import android.view.ContextMenu.ContextMenuInfo
import android.widget.Button
import android.widget.PopupMenu
import androidx.viewpager.widget.ViewPager
import com.google.android.material.tabs.TabLayout
import com.example.notes.DatePickerFragment
import com.example.notes.TimePickerFragment
import com.example.notes.Adapter.SectionPagerAdapter
import android.widget.TimePicker
import android.widget.TimePicker.OnTimeChangedListener
import androidx.fragment.app.Fragment

class NoteFragment : Fragment() {
    //  public String mListener;
    var title_was_changed = false
    private val text_color = 0
    private var note: Note? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (savedInstanceState != null) requireActivity().supportFragmentManager.popBackStack()
    }

    //Call onActivity Create method
    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        setHasOptionsMenu(true)
        return inflater.inflate(R.layout.fragment_note, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val arguments = arguments
        val time_date_alarm = view.findViewById<TextView>(R.id.time_date_alarm_view)
        val tvDescription = view.findViewById<TextView>(R.id.tvDescription)
        //  sharedPreferences = getActivity().getPreferences(Context.MODE_PRIVATE);
        title_was_changed = false
        if (arguments != null) {
            note = arguments.getParcelable(SELECTED_NOTE)
            if (note == null) note = Note.Companion.getNote(0)
            val tvTitle = view.findViewById<TextView>(R.id.tvTitle)
            tvTitle.text = note.getTitle()
            tvTitle.setTextColor(resources.getColor(MainActivity.note_text_color))
            tvTitle.addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(
                    charSequence: CharSequence,
                    i: Int,
                    i1: Int,
                    i2: Int
                ) {
                }

                override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                    val new_title = charSequence.toString()
                    note.setTitle(new_title)
                    if (note!!.getId(note) >= 0) {
                        CardSourceImpl.dataSource[note!!.getId(note)] = CardData(
                            new_title,
                            note.getDescription(), note.getCreationDate(),
                            note.getPictureID(), note!!.getId(note),
                            note.getLike()
                        )
                        val jsonCardDataAfterUpdate = GsonBuilder().create()
                            .toJson(CardSourceImpl.dataSource)
                        ListFragmentV2.sharedPreferences!!.edit()
                            .putString(ListFragmentV2.KEY, jsonCardDataAfterUpdate).apply()
                    }
                    title_was_changed = true
                    var rt = activity!!.findViewById<TextView>(note!!.getId(note))
                    if (rt == null) {
                        val k = 1
                        rt = activity!!.findViewById(k)
                    }
                    if (PDateTime == 0 && rt != null && note != null) rt.text = note.getTitle()
                }

                override fun afterTextChanged(editable: Editable) {}
            })
            time_date_alarm.text = note.getTimeDateAlarm()
            tvDescription.text = note.getDescription()
            tvDescription.setTextColor(resources.getColor(MainActivity.note_text_color))
            tvDescription.addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(
                    charSequence: CharSequence,
                    i: Int,
                    i1: Int,
                    i2: Int
                ) {
                }

                override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                    val new_description = charSequence.toString()
                    note.setDescription(new_description)
                    if (note!!.getId(note) >= 0) {
                        CardSourceImpl.dataSource[note!!.getId(note)] = CardData(
                            note.getTitle(),
                            new_description, note.getCreationDate(),
                            note.getPictureID(), note!!.getId(note),
                            note.getLike()
                        )
                        val jsonCardDataAfterUpdate = GsonBuilder()
                            .create().toJson(CardSourceImpl.dataSource)
                        ListFragmentV2.sharedPreferences!!.edit()
                            .putString(ListFragmentV2.KEY, jsonCardDataAfterUpdate).apply()
                    }
                }

                override fun afterTextChanged(editable: Editable) {}
            })
            time_date_alarm.setOnLongClickListener { view1: View? ->
                initDateTimePopupMenu(time_date_alarm)
                true
            }
            val buttonBack = view.findViewById<Button>(R.id.btnBack)
            buttonBack?.setOnClickListener { view1: View? ->
                if (title_was_changed) {
                    val rt = activity!!.findViewById<TextView>(note!!.getId(note))
                    rt.text = note.getTitle()
                }
                requireActivity().supportFragmentManager.popBackStack()
            }
        }
    }

    private fun initDateTimePopupMenu(view: TextView) {
        val activity: Activity = requireActivity()
        val popupMenu = PopupMenu(activity, view)
        activity.menuInflater.inflate(R.menu.popup_menu, popupMenu.menu)
        val item = popupMenu.menu.findItem(R.id.action_popup_delete)
        if (item != null) {
            item.isVisible = false
        }
        popupMenu.setOnMenuItemClickListener { menuItem -> // Слушатель диалога , сюда попадает управление
            if (menuItem.itemId == R.id.action_popup_correct) {
                val dialogDateTimeFragment: ParentDateTimeFragment =
                    ParentDateTimeFragment.Companion.newInstance(note)
                dialogDateTimeFragment.show(
                    requireActivity().supportFragmentManager,
                    "dialog_time_date_fragment"
                )
            }
            true
        }
        popupMenu.show()
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        val CurrentFragment = requireActivity().supportFragmentManager
            .findFragmentByTag("NOTE_FRAGMENT")
        if (!isLandscape && menu != null &&
            CurrentFragment != null && CurrentFragment.isVisible
        ) {
            menu.clear()
            inflater.inflate(R.menu.note_fragment_menu, menu)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        when (id) {
            R.id.send_note -> {
                // Сделать поиск
                createOneButtonAlertDialog("Сделать отослать заметку")
                return true
            }
            R.id.add_photo -> {
                createOneButtonAlertDialog("Сделать добавить фото")
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private val isLandscape: Boolean
        private get() = (resources.configuration.orientation
                == Configuration.ORIENTATION_LANDSCAPE)

    private fun createOneButtonAlertDialog(title_window: String) {
        val builder = AlertDialog.Builder(activity)
        builder.setTitle(title_window)
            .setMessage(R.string.version)
            .setPositiveButton(R.string.ok) { dialog, which -> }
        val d: Dialog = builder.show()
        var textViewId = d.context.resources.getIdentifier("android:id/alertTitle", null, null)
        var tv = d.findViewById<TextView>(textViewId)
        tv.setTextColor(resources.getColor(R.color.set_text_toast))
        tv.setBackgroundColor(resources.getColor(R.color.teal_200))
        textViewId = d.context.resources.getIdentifier("android:id/message", null, null)
        tv = d.findViewById(textViewId)
        tv.setTextColor(resources.getColor(android.R.color.holo_red_dark))
        textViewId = d.context.resources.getIdentifier("android:id/button1", null, null)
        val b = d.findViewById<Button>(textViewId)
        b.setTextColor(resources.getColor(android.R.color.holo_green_dark))
        b.setBackgroundColor(resources.getColor(R.color.color_time_date))
    }

    companion object {
        //  private OnDialogListener ColorListener;
        const val SELECTED_NOTE = "note"
        var PDateTime = 0
        fun newInstance(note: Note?): NoteFragment {
            val fragment = NoteFragment()
            val args = Bundle()
            args.putParcelable(SELECTED_NOTE, note)
            fragment.arguments = args
            return fragment
        }
    }
}